test/suites/integration/drift_test.go

```diff
@@ -0,0 +1,622 @@
+			g.Expect(env.Client.Get(env.Context, client.ObjectKeyFromObject(nodeClaim), nodeClaim)).To(Succeed())
+
+			g.Expect(nodePool.Annotations).To(HaveKeyWithValue(v1.NodePoolHashAnnotationKey, expectedHash))
+			g.Expect(nodePool.Annotations).To(HaveKeyWithValue(v1.NodePoolHashVersionAnnotationKey, v1.NodePoolHashVersion))
+			g.Expect(nodeClaim.Annotations).To(HaveKeyWithValue(v1.NodePoolHashAnnotationKey, expectedHash))
+			g.Expect(nodeClaim.Annotations).To(HaveKeyWithValue(v1.NodePoolHashVersionAnnotationKey, v1.NodePoolHashVersion))
+		}).WithTimeout(30 * time.Second).Should(Succeed())
+
+		nodePool.Annotations = lo.Assign(nodePool.Annotations, map[string]string{
+			v1.NodePoolHashAnnotationKey:        "test-hash-1",
+			v1.NodePoolHashVersionAnnotationKey: "test-hash-version-1",
+		})
+		// Updating `nodePool.Spec.Template.Annotations` would normally trigger drift on all nodeclaims owned by the
+		// nodepool. However, the nodepool-hash-version does not match the controller hash version, so we will see that
+		// none of the nodeclaims will be drifted and all nodeclaims will have an updated `nodepool-hash` and `nodepool-hash-version` annotation
+		nodePool.Spec.Template.Annotations = lo.Assign(nodePool.Spec.Template.Annotations, map[string]string{
+			"test-key": "test-value",
+		})
+		nodeClaim.Annotations = lo.Assign(nodePool.Annotations, map[string]string{
+			v1.NodePoolHashAnnotationKey:        "test-hash-2",
+			v1.NodePoolHashVersionAnnotationKey: "test-hash-version-2",
+		})
+
+		// The nodeclaim will need to be updated first, as the hash controller will only be triggered on changes to the nodepool
+		env.ExpectUpdated(nodeClaim, nodePool)
+		expectedHash = nodePool.Hash()
+
+		// Expect all nodeclaims not to be drifted and contain an updated `nodepool-hash` and `nodepool-hash-version` annotation
+		Eventually(func(g Gomega) {
+			g.Expect(env.Client.Get(env.Context, client.ObjectKeyFromObject(nodePool), nodePool)).To(Succeed())
+			g.Expect(env.Client.Get(env.Context, client.ObjectKeyFromObject(nodeClaim), nodeClaim)).To(Succeed())
+
+			g.Expect(nodePool.Annotations).To(HaveKeyWithValue(v1.NodePoolHashAnnotationKey, expectedHash))
+			g.Expect(nodePool.Annotations).To(HaveKeyWithValue(v1.NodePoolHashVersionAnnotationKey, v1.NodePoolHashVersion))
+			g.Expect(nodeClaim.Annotations).To(HaveKeyWithValue(v1.NodePoolHashAnnotationKey, expectedHash))
+			g.Expect(nodeClaim.Annotations).To(HaveKeyWithValue(v1.NodePoolHashVersionAnnotationKey, v1.NodePoolHashVersion))
+		})
+	})
+	Context("Failure", func() {
+		It("should not disrupt a drifted node if the replacement node never registers", func() {
+			version, err := env.KubeClient.Discovery().ServerVersion()
+			Expect(err).To(BeNil())
+			if version.Minor < "28" {
+				Skip("This test is only valid for K8s >= 1.28")
+			}
+			admissionspolicy := &admissionregistrationv1.ValidatingAdmissionPolicy{
```

Comment:
Admissions policy are used to mock node registration failures.

Reply:
I was updating the status condition of the nodeclaim at first and that would reduce the time to less then a min: https://github.com/kubernetes-sigs/karpenter/pull/2193/files#diff-ec78a380674cc833d5255f7b850e5ae25337cd6099fed376cc7e177e449b19c8R216. However this felt more natural to me. We can go either way.

Reply:
? There is a registration delay on the KwokNodeClass -- we can't use that?

Reply:
I did consider using  the registration delay. However, as a tanant , I want to make sure these can be imported by any Karpenter cloud provider and run against any nodeclass. Using NodeClass specific logic would break that, like registration delay, and I have stayed away from solutions that can't work for all Cloud providers